virtual_rainforest/animals/dummy_animal_module.py

```diff
@@ -0,0 +1,93 @@
+
+# to do
+# - rework dispersal
+# - send portion of dead to carcass pool
+
+# current simplifications
+# - only herbivory (want: carnivory and omnivory)
+# - only endothermy (want: ectothermy)
+# - only iteroparity (want: semelparity)
+# - no development
+
+# notes to self
+# assume each grid = 1 km2
+# assume each tick = 1 day (28800s)
+# damuth ~ 4.23*mass**(-3/4) indiv / km2
+# waste_energy pool likely unnecessary
+#   better to excrete directly to external pools
+# only elephants disperse atm
+
+
+class Plant:
+    """This is a dummy class of plant cohorts for testing the animal module."""
+
+    def __init__(self, name: str, mass: float, position: int) -> None:
+        """The constructor for Plant class."""
+        self.name = name
+        """The functional-type name of the plant cohort."""
+        self.mass = mass
+        """The mass of the plant cohort [g]."""
+        self.energy = mass * 100
+        """The amount of energy in the plant cohort [j] [toy]."""
+        self.alive = "alive"
+        """Whether the cohort is alive [True] or dead [False]."""
+        self.energy_max = mass * 100
+        """The maximum amount of energy that the cohort can have [j] [toy]."""
+        self.position = position
+        """The grid location of the cohort [0-8]."""
+
+    def plant_growth(self) -> None:
+        """The function to logistically modify cohort energy to the energy_max value.
+
+        Args:
+            None: Toy implementation of growth is only a function of the
+                  current energy state.
+
+        Returns:
+            Modified value of self.energy
+        """
+        self.energy += 1 * self.energy * (1 - self.energy / self.energy_max)
+
+    def plant_death(self) -> str:
+        """The function to change the self.alive state from True to False.
+
+        Parameters:
+            None: Toy implementation of death is only a function of the
+                    current aliveness state.
+
+        Returns:
+            Modified value of self.alive.
+            An alert (str) informing you the cohort has died.
+        """
+        self.alive = "dead"
+        return f"""A {self.name} cohort died"""
+
+
+class SoilPool:
+    """This is a dummy class of soil pools for testing the animal module."""
+
+    def __init__(self, energy: float, position: int) -> None:
+        """The constructor for Soil class."""
+        self.energy = energy
+        """The amount of energy in the soil pool [j]."""
+        self.position = position
+        """The grid position of the soil pool [0-8]."""
+
+
+class CarcassPool:
+    """This is a class of carcass pools."""
+
+    def __init__(self, energy: float, position: int) -> None:
+        """The constructor for Carcass class."""
+        self.energy = energy
+        """The amount of energy in the carcass pool [j]."""
+        self.position = position
+        """The grid position of carcass pool pool [0-8]."""
```

Comment:
Is it `[0-8]` because your toy example is using a 3x3 grid? I wouldn't tie down the input values here even in the toy.

More generally, you're going to want a `CarcassPool` for each grid cell, so I would lean towards having a list of them `carcasses = [<CarcassPool>, <CarcassPool>, ..., <CarcassPool>]`. If those are in `cell_id` order then you can get the pool for cell_id `n` by getting `carcasses[n]`. That removes the need to encode `position` in the `CarcassPool` object. 

You could do it anyway, for information, but I wouldn't use it as a mechanism for finding the right one for a cell.

Reply:
Compare:

```python
for cell_index in cell_ids:
    this_carcass = carcasses[cell_index]
```

```python
for cell_index in cell_ids:
    this_carcass = [cp for cp in carcasses if cp.position == cell_index][0] 
```

Reply:
I was using 0-8 for a toy 3x3 grid. I intended that [0-8] comment as a note-to-self but I need to get out of the habit of thinking about comments that way. 

I have a list of CarcassPool in the, yet unpulled, AnimalModel class but I haven't been using the ordering for anything yet. I will try to build towards using that ordered structure for tracking.

 I would like to keep a position attribute on everything for now so I can keep a close eye on how dispersal and positional generation are working (or not working). It's less of an issue for CarcassPool but I anticipate things getting hairy for AnimalCohorts and I'd like to keep the tracking consistent. Does that make sense?

Reply:
Yup. From plants, it makes absolute sense to have a single `PlantCommunity` object assigned to each cell (until we do the `Triffid` extension at least). But for animal cohorts, it is probably much easier to have a single cohort with a spatial map of where the individuals are, rather than constantly having to divide that cohort onto the cell-specific impacts. So then `AnimalCohort` definitely needs a structure like:

```
AnimalCohort.spatial = {
    cell_id = [2,3,4,5,6],
    proportion = [0.1, 0.1, 0.6, 0.1, 0.1]
}
```
that records the cells being used by each cohort and the relative impacts in terms of energy extraction (waves hands vaguely) to allow core and peripheral range use.

Or something.

Reply:
That makes sense to me. I'll get a basic spatial distribution structure  in there for animal cohorts before I pull it.